File: Lesson.02 - Design Patterns - Creational/FactoryMethod/src/main/java/com/github/ASE/FactoryMethod/MarkDown/DocumentGenerator.java
package com.github.ASE.FactoryMethod.MarkDown;

import com.github.ASE.FactoryMethod.Footer;
import com.github.ASE.FactoryMethod.Header;
import com.github.ASE.FactoryMethod.Paragraph;
import com.github.ASE.FactoryMethod.Table;

public class DocumentGenerator extends com.github.ASE.FactoryMethod.DocumentGenerator {

    @Override
    public Header createHeader() {
        return new Header() {
            @Override
            public String getContent() {
                return "# Document Title";
            }
        };
    }

    @Override
    public Paragraph createParagraph() {
        return new Paragraph() {
            @Override
            public String getContent() {
                return "This is a Markdown paragraph.";
            }
        };
    }

    @Override
    public Table createTable() {
        return new Table() {
            @Override
            public String getContent() {
                return "| Name | Age |\n|------|-----|\n| John | 30  |";
            }
        };
    }

    @Override
    public Footer createFooter() {
        return new Footer() {
            @Override
            public String getContent() {
                return "---\n*End of Document*";
            }
        };
    }

}

File: Lesson.02 - Design Patterns - Creational/Python/FactoryMethod/markdown/document_generator.py
from document_generator import DocumentGenerator
from components.header import Header
from components.paragraph import Paragraph
from components.table import Table
from components.footer import Footer


class MarkdownDocumentGenerator(DocumentGenerator):
    def create_header(self) -> Header:
        return MarkdownHeader()

    def create_paragraph(self) -> Paragraph:
        return MarkdownParagraph()

    def create_table(self) -> Table:
        return MarkdownTable()

    def create_footer(self) -> Footer:
        return MarkdownFooter()


class MarkdownHeader(Header):
    def get_content(self) -> str:
        return "# Document Title"


class MarkdownParagraph(Paragraph):
    def get_content(self) -> str:
        return "This is a Markdown paragraph."


class MarkdownTable(Table):
    def get_content(self) -> str:
        return "| Name | Age |\n|------|-----|\n| John | 30  |"


class MarkdownFooter(Footer):
    def get_content(self) -> str:
        return "---\n*End of Document*"
